### PR TITLE
feat: update diagnostics and parsing interfaces BREAKING [SL-2114]

### PR DESCRIPTION
**Pull request by P0lip**
Should we rename `validations.ts` to `diagnostics.ts`?

### CODE REVIEW

File: src/parsers.ts [hunk 1]
```diff
@@ -1,25 +1,61 @@
...
+
+  /**
+   * The end position. It is after or equal to start.
+   */
+  readonly end: IPosition;
 }
 
-export type IParserResultPointers = Dictionary<IPathPosition, string>;
+export interface ILocation {
+  uri: DocumentUri;
```
Comment (marbemac) [inline src/parsers.ts]: Should we make this optional?

File: src/parsers.ts [hunk 2]
```diff
@@ -1,25 +1,61 @@
...
-export type ObjPath = Array<string | number>;
+export type DocumentUri = string;
+export type Segment = string | number;
+export type JSONPath = Segment[];
+
+export interface IParserResult<T = any> {
+  data: T;
+  diagnostics: IDiagnostic[];
+
+  // One of these, not sure? jsonc uses offset, language server spec uses line + character :/
```
Comment (marbemac) [inline src/parsers.ts]: If we've decided, please remove the open question.

File: src/validations.ts [hunk 3]
```diff
@@ -1,39 +1,87 @@
...
 }
 
-export interface IValidationResult {
-  /** Name identifier for this type of validation. */
-  name: string;
+export type DiagnosticTag = string;
 
-  /** Validation severity as a number. */
-  severity: ValidationSeverity;
+export interface IDiagnostic {
```
Comment (marbemac) [inline src/validations.ts]: I think we should add optional `path?: JSONPath`.

File: src/parsers.ts (hunk 1 above)
Comment (P0lip) [reply src/parsers.ts]: @marbemac Yeah, I never actually provide any value at the moment, so making it optional makes sense.

**Pull request merged**